delta/service/src/main/scala/ch/epfl/bluebrain/nexus/delta/service/acls/AclsImpl.scala

```diff
@@ -0,0 +1,149 @@
+import ch.epfl.bluebrain.nexus.delta.service.cache.{KeyValueStore, KeyValueStoreConfig}
+import ch.epfl.bluebrain.nexus.sourcing.processor.{AggregateConfig, ShardedAggregate, StopStrategy}
+import ch.epfl.bluebrain.nexus.sourcing._
+import monix.bio.{IO, UIO}
+
+final class AclsImpl private (agg: AclsAggregate, eventLog: EventLog[Envelope[AclEvent]], index: AclsCache)
+    extends Acls {
+
+  override def fetch(address: AclAddress): UIO[Option[AclResource]] = agg.state(address.string).map(_.toResource)
+
+  override def fetchAt(address: AclAddress, rev: Long): IO[AclRejection.RevisionNotFound, Option[AclResource]] =
+    if (rev == 0L) UIO.pure(None)
+    else
+      eventLog
+        .currentEventsByPersistenceId(
+          s"$entityType-${URLEncoder.encode(address.string, "UTF-8")}",
+          Long.MinValue,
+          Long.MaxValue
+        )
+        .takeWhile(_.event.rev <= rev)
+        .fold[AclState](AclState.Initial) {
+          case (state, event) => Acls.next(state, event.event)
+        }
+        .compile
+        .last
+        .hideErrors
+        .flatMap {
+          case Some(state) if state.rev == rev => UIO.pure(state.toResource)
+          case Some(_)                         =>
+            fetch(address).flatMap(res => IO.raiseError(RevisionNotFound(rev, res.map(_.rev).getOrElse(0L))))
+          case None                            => IO.raiseError(RevisionNotFound(rev, 0L))
+        }
+
+  override def list(filter: AclAddressFilter): UIO[AclCollection] = {
+    index.values.map(as => AclCollection(as.toSeq: _*).fetch(filter))
+
+  }
+
+  override def listSelf(filter: AclAddressFilter)(implicit caller: Caller): UIO[AclCollection] =
+    list(filter).map(_.filter(caller.identities))
+
+  override def replace(address: AclAddress, acl: Acl, rev: Long)(implicit
+      caller: Identity.Subject
+  ): IO[AclRejection, AclResource] = eval(ReplaceAcl(address, acl, rev, caller))
+
+  override def append(address: AclAddress, acl: Acl, rev: Long)(implicit
+      caller: Identity.Subject
+  ): IO[AclRejection, AclResource] = eval(AppendAcl(address, acl, rev, caller))
+
+  override def subtract(address: AclAddress, acl: Acl, rev: Long)(implicit
+      caller: Identity.Subject
+  ): IO[AclRejection, AclResource] = eval(SubtractAcl(address, acl, rev, caller))
+
+  override def delete(address: AclAddress, rev: Long)(implicit
+      caller: Identity.Subject
+  ): IO[AclRejection, AclResource] = eval(DeleteAcl(address, rev, caller))
+
+  private def eval(cmd: AclCommand): IO[AclRejection, AclResource] =
+    for {
+      evaluationResult <- agg.evaluate(cmd.address.string, cmd).mapError(_.value)
+      resource         <- IO.fromOption(
+                            evaluationResult.state.toResource,
+                            UnexpectedInitialState(cmd.address)
+                          )
+      _                <- index.put(cmd.address, resource)
```

Comment:
So are we populating the cache synchronously here instead of how we did it before reading events from the primary store asynchronously?

It could be a good option but worth it to discussion. Maybe tomorrow morning meeting?

Reply:
That's what we do currently as well. We update the cache on change to propagate the changes faster, but we also update it from primary store events to rebuild the cache in case of restarts.

Reply:
Ah ok. So that's the exception.
So where is the code that populates the cache from the primary store events? I guess it should be present on the method:
```scala
 private def cache(aclsConfig: AclsConfig)(implicit as: ActorSystem[Nothing]): AclsCache = ???
```

That, or making the `AclsCache` a class/object and do the populate of acl resources there.

Reply:
Populating the cache incrementally from calling fetch could also be an option no ?
So that only the "hot" acls find themselves in cache.

Reply:
I think we need to have all the ACLs in the cache, since we use the cache for listings.